src/component/Home.js

```diff
@@ -1,11 +1,13 @@
-import React from "react";
+import React, { useState } from "react";
 import GetToken from './GetToken';
 
-function Home() {
+function Home(props) {
+  const { token: [token, setToken] } = { token: useState(null), ...(props.state || {}) };
```

Comment:
hey @mindyzwan can you walk me through what is happening inside this component with state / props? I looked through the information Alejandro sent us about lifting state up but I'm getting lost in this part of the code.

Reply:
Sure! And I don't fully understand it myself yet, so figuring out how to explain it will probably be good for me! I'll start by sharing this link, which is where I started to understand what was going on with lifting state: https://codesandbox.io/s/oq587qpno9?from-embed

Let me see if I can find more resources and try to explain my thought process...

Reply:
I think the basic idea is: we set the `token` state in one of the 'parent' components, and then pass that state to whatever other components might use that state via props. 

In `App.js`, the only component we have is the Router component, so you can pass that state down by doing this: `<RouterComponent state={{ token: [token, setToken] }}/>;` (line 11 of the `App.js` file). So now, using the declaration on line 11 of `Router.js`, we can set the `token` state in `Router` by referencing the props we passed in. 

We use the idea of a token or having/not having a token in a couple of other components - `GetToken` and `JoinList`- so we basically repeat that process and pass the state down from `Router` to those other components. BUT `GetToken` is only accessible through `Home` so we have to pass it through `Home` to get to `GetToken`. 

I feel like this approach is fairly messy - which makes me wonder if there's another way to do it. I'd be really curious if this approach is what @alejandronanez was referencing in his comment.

Reply:
I feel that I'm 50% of the way to understanding, but I don't get this line, which is repeated a few times:

`const { token: [token, setToken] } = { token: useState(null), ...(props.state || {}) };
`
Is this creating a constant that is named { token: [token, setToken] }? I don't get that at all, and couldn't find that in any tutorials to explain it to me.

Reply:
I *think* that's what's taking the props and turning it into state for each component it's in. In the link Alejandro sent (https://reactjs.org/docs/lifting-state-up.html), they use `this.props.something`, but I'm thinking the syntax here is more hooks-like? Kinda fuzzy on it...there are some fundamental understandings of props and state that I haven't gotten to yet in my studies, so I don't fully get what's happening in that line.

Reply:
I'm not super familiar with hooks, so I'm not the best to comment on this implementation. My hunch is that this is a bit more complicated than necessary to accomplish the task of lifting state. Remember that we should strive for the simplest solution that satisfies the requirements because then we have only the _necessary_ complexity in our app.

Because I haven't worked much with hooks, I'll let @alejandronanez give guidance on the specific implementation details.

Reply:
Hey there :)

Well, as @segdeha said, this is getting a bit more complicated than it should 😅:

1. Is not a good practice to have a prop called `state`, think of `state` as a reserved word in React.
2. From what I can see, you got the idea of what lifting state up means, I'd clean up the way you pass down the props from App to its children.

I created a Codesandbox with an example very similar to what you're doing and added a bunch of comments there, please, let me know if there's anything confusing after reading my code! https://codesandbox.io/s/tcl-lifting-state-up-5mzkg

In bigger applications, you tend to use `Context` to avoid _prop drilling_, which is passing down props 2+ levels deep like this:

```js
function App() {
  const name = "alejandro";

 return (
    <ComponentOne name={name} />
 )
}

function ComponentOne({ name }) {
 return (
    <ComponentTwo name={name} />
 )
}

function ComponentTwo({ name }) {
 return (
    <ComponentThreeInterestedInName name={name} />
 )
}

function ComponentThreeInterestedInName({ name }) {
 return (
    <h1>Hello {name}</h1>
 )
}
```

If you don't choose to use `Context`, you can use composition instead (and it's pretty cool!)

Here's a video about using Composition instead of Context

https://www.youtube.com/watch?v=3XaXKiXtNjw&t=5s

I highly recommend you watch this video: @stacietaylorcima @mindyzwan @michellejae @ClaudiaStar @gunderodd